### PR TITLE
Shared model/new block loader

### PR DESCRIPTION
**Pull request by x3medima17**
### Description of the Change
Migrate `BlockLoader` and `BlockLoaderService` to the new model

### Benefits
No dependency on old model

### Possible Drawbacks 
None

### CODE REVIEW

**Comment by l4l** [comment]
Is the pr up-to-date?

File: irohad/network/impl/block_loader_impl.cpp [hunk 1]
```diff
@@ -77,18 +75,18 @@ rxcpp::observable<std::shared_ptr<Block>> BlockLoaderImpl::retrieveBlocks(
...
+            this->getPeerStub(**peer).retrieveBlocks(&context, request);
         while (reader->Read(&block)) {
           shared_model::proto::TransportBuilder<
               shared_model::proto::Block,
               shared_model::validation::DefaultSignableBlockValidator>()
               .build(block)
               .match(
                   // success case
-                  [&subscriber](
+                  [this, &context, &subscriber](
```
Comment (Alexey-N-Chernyshov) [inline irohad/network/impl/block_loader_impl.cpp]: this, context is not needed here

File: irohad/network/impl/block_loader_impl.cpp [hunk 2]
```diff
@@ -20,6 +20,7 @@
 #include <algorithm>
 
 #include <grpc++/create_channel.h>
+#include <algorithm>
```
Comment (Alexey-N-Chernyshov) [inline irohad/network/impl/block_loader_impl.cpp]: This is duplication. Please remove.

File: irohad/network/impl/block_loader_impl.hpp [hunk 3]
```diff
@@ -26,6 +26,7 @@
 #include "ametsuchi/peer_query.hpp"
 #include "loader.grpc.pb.h"
 #include "logger/logger.hpp"
+#include "validators/block_validator.hpp"
```
Comment (Alexey-N-Chernyshov) [inline irohad/network/impl/block_loader_impl.hpp]: default_validators.hpp should be enough, please remove it

File: irohad/network/impl/block_loader_service.cpp [hunk 4]
```diff
@@ -48,23 +47,21 @@ grpc::Status BlockLoaderService::retrieveBlock(
     ::grpc::ServerContext *context,
     const proto::BlockRequest *request,
     protocol::Block *response) {
-  const auto hash = stringToBlob<Block::HashType::size()>(request->hash());
-  if (not hash) {
-    log_->error("Bad hash in request, {}",
-                bytestringToHexstring(request->hash()));
+
```
Comment (Alexey-N-Chernyshov) [inline irohad/network/impl/block_loader_service.cpp]: extra space, remove

File: test/module/irohad/model/model_mocks.hpp [hunk 5]
```diff
@@ -22,11 +22,22 @@
 #include "ametsuchi/wsv_command.hpp"
 #include "ametsuchi/wsv_query.hpp"
 #include "model/command.hpp"
+#include "model/model_crypto_provider.hpp"
 #include "model/query_execution.hpp"
 
 namespace iroha {
   namespace model {
 
+    class MockCryptoProvider : public ModelCryptoProvider {
```
Comment (Alexey-N-Chernyshov) [inline test/module/irohad/model/model_mocks.hpp]: ModelCryptoProvider is deprecated class and now it is used only in CLI. I don't know why this mock could be needed for.
I think changes in this file should be reverted.

File: test/module/irohad/network/block_loader_test.cpp [hunk 6]
```diff
@@ -31,6 +31,7 @@
 #include "datetime/time.hpp"
 #include "framework/test_subscriber.hpp"
 #include "module/irohad/ametsuchi/ametsuchi_mocks.hpp"
+#include "module/irohad/model/model_mocks.hpp"
```
Comment (Alexey-N-Chernyshov) [inline test/module/irohad/network/block_loader_test.cpp]: unnecessary include

File: irohad/network/impl/block_loader_service.cpp [hunk 7]
```diff
@@ -36,8 +36,7 @@ grpc::Status BlockLoaderService::retrieveBlocks(
     ::grpc::ServerWriter<::iroha::protocol::Block> *writer) {
   storage_->getBlocksFrom(request->height())
       .map([this](auto block) {
```
Comment (Alexey-N-Chernyshov) [inline irohad/network/impl/block_loader_service.cpp]: [this] in capture is not used

File: irohad/network/impl/block_loader_service.cpp [hunk 8]
```diff
@@ -48,23 +47,21 @@ grpc::Status BlockLoaderService::retrieveBlock(
...
   }
 
   boost::optional<protocol::Block> result;
   storage_->getBlocksFrom(1)
       .filter([hash](auto block) {
-        return shared_model::crypto::toBinaryString(block->hash())
-            == hash->to_string();
+        return block->hash() == hash;
       })
       .map([this](auto block) {
```
Comment (Alexey-N-Chernyshov) [inline irohad/network/impl/block_loader_service.cpp]: [this] in capture is not used

File: test/module/irohad/network/block_loader_test.cpp [hunk 9]
```diff
@@ -203,7 +204,7 @@ TEST_F(BlockLoaderTest, ValidWhenBlockPresent) {
   auto block = loader->retrieveBlock(peer_key, requested.hash());
 
   ASSERT_TRUE(block);
-  ASSERT_EQ(*(*block), requested);
+  ASSERT_EQ(*(*block).operator->(), requested);
```
Comment (Alexey-N-Chernyshov) [inline test/module/irohad/network/block_loader_test.cpp]: just `**block`

File: irohad/network/impl/block_loader_service.cpp [hunk 10]
```diff
@@ -48,23 +47,19 @@ grpc::Status BlockLoaderService::retrieveBlock(
...
   boost::optional<protocol::Block> result;
   storage_->getBlocksFrom(1)
-      .filter([hash](auto block) {
-        return shared_model::crypto::toBinaryString(block->hash())
-            == hash->to_string();
-      })
-      .map([this](auto block) {
-        return factory_.serialize(
-            *std::unique_ptr<iroha::model::Block>(block->makeOldModel()));
+      .filter([hash](auto block) { return block->hash() == hash; })
```
Comment (l4l) [inline irohad/network/impl/block_loader_service.cpp]: Capture hash by reference

File: irohad/network/impl/block_loader_service.cpp [hunk 11]
```diff
@@ -35,9 +34,9 @@ grpc::Status BlockLoaderService::retrieveBlocks(
     const proto::BlocksRequest *request,
     ::grpc::ServerWriter<::iroha::protocol::Block> *writer) {
   storage_->getBlocksFrom(request->height())
-      .map([this](auto block) {
-        return factory_.serialize(
-            *std::unique_ptr<iroha::model::Block>(block->makeOldModel()));
+      .map([](auto block) {
+        return std::static_pointer_cast<shared_model::proto::Block>(block)
```
Comment (l4l) [inline irohad/network/impl/block_loader_service.cpp]: Maybe better to add an assertion, for cast checking?

File: irohad/network/impl/block_loader_service.cpp [hunk 12]
```diff
@@ -16,8 +16,7 @@
  */
 
 #include "network/impl/block_loader_service.hpp"
-
-#include "common/byteutils.hpp"
+#include "backend/protobuf/block.hpp"
 
 using namespace iroha;
```
Comment (l4l) [inline irohad/network/impl/block_loader_service.cpp]: Remove redundant using-namespaces plz (iroha::model and maybe smth else)

File: irohad/network/impl/block_loader_impl.cpp [hunk 13]
```diff
@@ -17,9 +17,8 @@
 
 #include "network/impl/block_loader_impl.hpp"
 
-#include <algorithm>
-
 #include <grpc++/create_channel.h>
+#include <algorithm>
```
Comment (Alexey-N-Chernyshov) [inline irohad/network/impl/block_loader_impl.cpp]: order is broken

File: irohad/network/impl/block_loader_service.cpp (hunk 11 above)
Comment (x3medima17) [reply irohad/network/impl/block_loader_service.cpp]: How  about `dynamic_cast`?

**Pull request merged**